mirrord-layer/src/sockets.rs

```diff
@@ -1,240 +1,361 @@
-        let (write_socket, read_socket) = socketpair_stream().unwrap();
-        let read_fd = read_socket.as_raw_fd();
-        let socket = Socket {
-            read_fd,
-            read_socket,
-            write_socket,
-        };
-
-        self.new_sockets.lock().unwrap().insert(read_fd, socket);
+#[derive(Debug)]
+#[allow(dead_code)]
+struct Socket {
+    fd: RawFd,
+    domain: c_int,
+    type_: c_int,
+    protocol: c_int,
+    pub state: SocketState,
+}
 
-        read_fd
+impl PartialEq for Socket {
+    fn eq(&self, other: &Self) -> bool {
+        self.fd == other.fd
     }
+}
 
-    pub fn convert_to_connection_socket(&self, sockfd: SockFd, address: SocketAddr) {
-        let mut sockets = self.new_sockets.lock().unwrap();
-
-        // let mut sockets = self.connections.lock().unwrap();
-        if let Some(socket) = sockets.remove(&sockfd) {
-            // socket.port = port;
-            self.connections.lock().unwrap().insert(
-                sockfd,
-                address.port(),
-                ConnectionSocket {
-                    read_fd: socket.read_fd,
-                    read_socket: socket.read_socket,
-                    write_socket: socket.write_socket,
-                    address,
-                    state: ConnectionState::Bound,
-                },
-            );
-        } else {
-            error!("No socket found for fd: {}", sockfd);
-        }
-    }
+impl Eq for Socket {}
 
-    pub fn set_connection_state(&self, sockfd: SockFd, state: ConnectionState) -> Result<(), ()> {
-        let mut connections = self.connections.lock().unwrap();
-        if let Some(connection) = connections.get_mut(&sockfd) {
-            if connection.state != state {
-                connection.state = state;
-                Ok(())
-            } else {
-                error!("No connection found for fd: {}", sockfd);
-                Err(())
-            }
-        } else {
-            error!("No connection found for fd: {}", sockfd);
-            Err(())
-        }
+impl Hash for Socket {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.fd.hash(state);
     }
+}
 
-    pub fn get_connection_socket_address(&self, sockfd: SockFd) -> Option<SocketAddr> {
-        let sockets = self.connections.lock().unwrap();
-        sockets.get(&sockfd).map(|socket| socket.address)
+impl Borrow<RawFd> for Socket {
+    fn borrow(&self) -> &RawFd {
+        &self.fd
     }
+}
+
+#[derive(Default)]
+struct State {
+    sockets: HashSet<Socket>,
+}
 
-    pub fn get_data_socket_address(&self, sockfd: SockFd) -> Option<SocketAddr> {
-        let sockets = self.data.lock().unwrap();
-        sockets.get(&sockfd).map(|socket| socket.address)
+lazy_static! {
+    static ref STATE: Mutex<State> = Mutex::new(State::default());
+}
+
+#[inline]
+fn is_ignored_port(port: Port) -> bool {
+    port == 0 || (port > 50000 && port < 60000)
+}
+
+fn socket(domain: c_int, type_: c_int, protocol: c_int) -> RawFd {
```

Comment:
I'm going to add my interpretation of what the hooks do, let me know if I didn't get it right, and if I did please add it as function documentation.
For socket:
Create the socket, add it to State, and return it to the calling application.